### PR TITLE
nix-daemon: add support for 'idle' IO scheduling class

### PR DESCRIPTION
**Pull request by jerith666**
###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/nix-daemon.nix [hunk 1]
```diff
@@ -151,11 +151,13 @@ in
       };
 
       daemonIONiceLevel = mkOption {
-        type = types.int;
+        type = types.either types.int (types.enum ["idle"]);
```
Comment (peti) [inline nixos/modules/services/misc/nix-daemon.nix]: It feels like cramming this "idle" setting in here makes the user interface weird and counter-intuitive. Personally, I'd rather have an additional setting for this purpose which accepts a boolean to enable/disable "idle".

Comment (jerith666) [inline nixos/modules/services/misc/nix-daemon.nix]: I heard what you're saying.  The problem is that "idle" supersedes the numeric priority, so then the documentation for the numeric priority setting would have to have a clause like "... unless idle=true, in which case this setting is ignored" ... which I think is *also* confusing.  :-/

Comment (joachifm) [inline nixos/modules/services/misc/nix-daemon.nix]: What if we had "idle", "low" (best-effort, 7), "inherit" (whatever the OS determines), and "high" (best-effort, 0). Does anybody care about granular best-effort levels beyond lowest and highest?

Comment (jerith666) [inline nixos/modules/services/misc/nix-daemon.nix]: I like that idea, except that it's not backwards-compatible.  How would we deal with that issue?

Comment (joachifm) [inline nixos/modules/services/misc/nix-daemon.nix]: There's a pretty flexible system for dealing with that, see https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/rename.nix, it supports e.g., mapping old option declarations to new ones, establishing aliases, &c

Comment (jerith666) [inline nixos/modules/services/misc/nix-daemon.nix]: Ah, I think I did run across that at some point in the past.  Okay, I'll give it a look and rework this PR.  Thanks!

**Comment by edolstra** [comment]
To be honest, I feel we should remove options like `daemonIONiceLevel` (and certainly not make them more complicated), because they just alias systemd options in an ad hoc way. Systemd provides two dozen or so resource control directives, and we have lots of services, so that's hundreds of options we could potentially add. I think it's better to use the generic mechanism, e.g.
```
systemd.services.nix-daemon.serviceConfig.IOSchedulingClass = "idle";
```
because this is well-documented in the `systemd.resource-control` manpage, and generalizes to any systemd service, not just `nix-daemon`.

It's also debatable whether using the `idle` class is a good idea for `nix-daemon`. On a I/O-loaded system, doesn't this mean that builds might never finish?

**Comment by jerith666** [comment]
I definitely understand that perspective.  To play devil's advocate for a minute, though:

1) The `*.serviceConfig.*` options are not checked for correctness by nix, whereas purpose-built options like `daemonIONiceLevel` are.  It's much easier to discover `configuration.nix` typos etc. at `nix-build` time rather than by inspecting `systemctl status nix-daemon` output etc.

2) Most systemd options don't make sense for most services.  By "promoting" certain options like this, we help guide users to those that are most likely to make sense for each service.  (And limit our own maintenance overhead.)

WRT whether builds are starved: I've actually had the opposite problem, using `idle` has not had the effect I hoped for of making my desktop system more usable while builds are running in the background.  :-/

**Comment by joachifm** [comment]
@jerith666 I had assumed that the motivation for this was as you described; for many services there's definitely levels of commitment and it makes some sense to reflect that in the interface, but it's not clear how best to do it ...  An alternative approach is to document a general procedure for service resource control via systemd.

Fwiw, I've had more luck using cgroups to constrain Nix builds (giving fewer io slices, constraining memory, &c).

**Comment by jerith666** [comment]
Okay, thanks for the feedback everyone.  I'm going to withdraw this pull request for now, while I experiment a bit with the cgroups approach that @joachifm recommended and contemplate ways to address @edolstra's concerns.

**Pull request closed without merging**